runtime/interpreter/encode.go

```diff
@@ -280,339 +285,483 @@ func (e *Encoder) Encode(
+		return e.encodeUFix64(v)
 
 	// String
 
 	case *StringValue:
-		return e.prepareString(v), nil
+		return e.enc.EncodeString(v.Str)
 
 	// Collections
 
 	case *ArrayValue:
-		return e.prepareArray(v, path, deferrals)
+		return e.encodeArray(v, path, deferrals)
 
 	case *DictionaryValue:
-		return e.prepareDictionaryValue(v, path, deferrals)
+		return e.encodeDictionaryValue(v, path, deferrals)
 
 	// Composites
 
 	case *CompositeValue:
-		return e.prepareCompositeValue(v, path, deferrals)
+		return e.encodeCompositeValue(v, path, deferrals)
 
 	// Some
 
 	case *SomeValue:
-		return e.prepareSomeValue(v, path, deferrals)
+		return e.encodeSomeValue(v, path, deferrals)
 
 	// Storage
 
 	case PathValue:
-		return e.preparePathValue(v), nil
+		return e.encodePathValue(v)
 
 	case CapabilityValue:
-		return e.prepareCapabilityValue(v)
+		return e.encodeCapabilityValue(v)
 
 	case LinkValue:
-		return e.prepareLinkValue(v)
+		return e.encodeLinkValue(v)
 
 	// Type
 
 	case TypeValue:
-		return e.prepareTypeValue(v)
+		return e.encodeTypeValue(v)
 
 	default:
-		return nil, EncodingUnsupportedValueError{
+		return EncodingUnsupportedValueError{
 			Path:  path,
 			Value: v,
 		}
 	}
 }
 
-func (e *Encoder) prepareNil() interface{} {
-	return nil
+// cborVoidValue represents the CBOR value:
+//
+// 	cbor.Tag{
+// 		Number: cborTagVoidValue,
+// 		Content: nil
+// 	}
+//
+var cborVoidValue = []byte{
+	// tag
+	0xd8, cborTagVoidValue,
+	// null
+	0xf6,
 }
 
-func (e *Encoder) prepareVoid() cbor.Tag {
+// encodeVoid writes a value of type Void to the encoder
+//
+func (e *Encoder) encodeVoid() error {
 
 	// TODO: optimize: use 0xf7, but decoded by github.com/fxamacker/cbor/v2 as Go `nil`:
 	//   https://github.com/fxamacker/cbor/blob/a6ed6ff68e99cbb076997a08d19f03c453851555/README.md#limitations
 
-	return cbor.Tag{
-		Number: cborTagVoidValue,
-	}
-}
-
-func (e *Encoder) prepareBool(v BoolValue) bool {
-	return bool(v)
+	return e.enc.EncodeRawBytes(cborVoidValue)
 }
 
-func (e *Encoder) prepareInt(v IntValue) cbor.Tag {
-	return cbor.Tag{
-		Number:  cborTagIntValue,
-		Content: prepareBigInt(v.BigInt),
-	}
-}
-
-func (e *Encoder) prepareInt8(v Int8Value) cbor.Tag {
-	return cbor.Tag{
-		Number:  cborTagInt8Value,
-		Content: v,
+// encodeInt encodes IntValue as
+// cbor.Tag{
+//		Number:  cborTagIntValue,
+//		Content: *big.Int(v.BigInt),
+// }
+func (e *Encoder) encodeInt(v IntValue) error {
+	err := e.enc.EncodeRawBytes([]byte{
+		// tag number
+		0xd8, cborTagIntValue,
+	})
+	if err != nil {
+		return err
 	}
+	return e.enc.EncodeBigInt(v.BigInt)
 }
 
-func (e *Encoder) prepareInt16(v Int16Value) cbor.Tag {
-	return cbor.Tag{
-		Number:  cborTagInt16Value,
-		Content: v,
+// encodeInt8 encodes Int8Value as
+// cbor.Tag{
+//		Number:  cborTagInt8Value,
+//		Content: int8(v),
+// }
+func (e *Encoder) encodeInt8(v Int8Value) error {
+	err := e.enc.EncodeRawBytes([]byte{
+		// tag number
+		0xd8, cborTagInt8Value,
+	})
```

Comment:
Is there any performance difference when refactoring the argument into a global, like `cborVoidValue`? (same question for the other encoding functions)

Reply:
Good question!  It looks like there's not much performance difference when using global (on linux_amd64).  

One reason I didn't use global was because `[]byte{0xd8, cborTagInt8Value}` is incomplete CBOR while `cborVoidValue` is complete CBOR (valid CBOR).  Also confirmed the local slice doesn't escape to the heap.
```
BenchmarkGlobalSlice-4   	68186745	        16.4 ns/op	       8 B/op	       0 allocs/op
BenchmarkLocalSlice-4    	67143350	        16.3 ns/op	       8 B/op	       0 allocs/op
```

Reply:
Awesome, thanks for testing this 👍